python functionalization: support higher order ops

bdhirsh opened the pull request:
We now have two types of functionalization, C++ Functionalization (through the `Functionalize` dispatch key), and python functionalization (through the `FunctionalTensorMode` torch_dispatch mode).

This means that all higher order ops need custom functionalization rules for the python variant too. I added them here, as well as a helper function `dispatch_functionalize()` - equivalent to `torch.func.functionalize()`, except that it uses `FunctionalTensorMode`.

In theory we could have secretly switched `torch.func.functionalize` to use `FunctionalTensorMode`. This would be BC-breaking, though, since `FunctionalTensorMode` isn't composable with the other functorch transforms (the functorch layer-mode stack doesn't know how to re-order torch_dispatch modes arbitrarily).

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105236
* #104974
* #104483
* #106406
* #109662
* #109023
* #108654
* #109632
* __->__ #108656
* #109248
* #109024

bdhirsh committed:
[not read yet] python functionalization: support higher order ops

[ghstack-poisoned]

bdhirsh committed:
Update on "[not read yet] python functionalization: support higher order ops"




[ghstack-poisoned]

bdhirsh committed:
Update on "[not read yet] python functionalization: support higher order ops"




[ghstack-poisoned]

bdhirsh committed:
Update on "[not read yet] python functionalization: support higher order ops"




[ghstack-poisoned]

bdhirsh committed:
Update on "[not read yet] python functionalization: support higher order ops"




[ghstack-poisoned]

bdhirsh committed:
Update on "[not read yet] python functionalization: support higher order ops"




[ghstack-poisoned]

bdhirsh commented:
cc @zou3519, this is ready for review now (I think I've gotten all tests to pass, although I'll double check CI when it's done).

zou3519 commented on functorch/experimental/_map.py:
```diff
@@ -326,6 +345,24 @@ def map_fake_tensor_mode(f, num_mapped, *args):
     return map_dense(f, num_mapped, *args)
 
 
+@map_impl.py_impl(FunctionalTensorMode)
+def map_functional_tensor_mode(f, num_mapped, *args):
+    from torch._functorch.aot_autograd import from_fun, to_fun
+
+    xs = args[:num_mapped]
+    pos_args = args[num_mapped:]
+    unwrapped_xs = pytree.tree_map_only(FunctionalTensor, from_fun, xs)
+    unwrapped_args = pytree.tree_map_only(FunctionalTensor, from_fun, pos_args)
+
+    functional_map_fn = dispatch_functionalize(f)
+
+    with unset_functional_temporarily():
+        map_return = map_impl(
+            functional_map_fn, num_mapped, *unwrapped_xs, *unwrapped_args
+        )
+        return pytree.tree_map_only(torch.Tensor, to_fun, map_return)
+
```
Could we write a generic HigherOrderOp implementation for FunctionalTensorMode and register it by default to all HigherOrderOp? It sounds like for ATen operators:
- FunctionalTensorMode unwraps the FunctionalTensor
- then does a redispatch
- the actual functionalization kernel is handled at DispatchKey::Functionalize

So to be consistent, we'd want the same for HigherOrderOp. There are two approaches we could do here:
- just have HigherOrderOp's FunctionalTensorMode implementatation call `FunctionalTensorMode.__torch_dispatch__(higherorderop, args, kwargs)`. (Might not work too well, since `__torch_dispatch__` expects OpOverload)
- Write some code that unwraps all the inputs, does the TLS-y stuff, and then does a redispatch

zou3519 replied:
Why isn't the map Functional Tensor rule just the following:
```

@map_impl.py_impl(FunctionalTensorMode)
def map_functional_tensor_mode(f, num_mapped, *args):
    unwrapped = pytree.tree_map_only(FunctionalTensor, from_fun, args)
    with unset_functional_temporarily():
        map_return = map_impl(f, num_mapped, unwrapped)
    return pytree.tree_map_only(torch.Tensor, to_fun, map_return)
```

bdhirsh replied:
hmm I think you're right - I'll update it to that.

It currently looks slightly different:
```
@map_impl.py_impl(FunctionalTensorMode)
def map_functional_tensor_mode(f, num_mapped, *args):
    xs = args[:num_mapped]
    pos_args = args[num_mapped:]
    unwrapped_xs = pytree.tree_map_only(FunctionalTensor, from_fun, xs)
    unwrapped_args = pytree.tree_map_only(FunctionalTensor, from_fun, pos_args)

    functional_map_fn = dispatch_functionalize(f)

    with unset_functional_temporarily():
        map_return = map_impl(
            functional_map_fn, num_mapped, *unwrapped_xs, *unwrapped_args
        )
        return pytree.tree_map_only(torch.Tensor, to_fun, map_return)
```

I separated args into xs and pos_args to copy what the `Functionalize` kernel does - but that's probably unnecessary, since we know that the Functionalize kernel will handle it.

zou3519 replied:
If we can update the map FunctionalTensor rule to that, we can also update the cond FunctionalTensor rule to that, right? If so then we can have a single generic FunctionalTensor implementation for HigherOrderOp

bdhirsh replied:
`cond` is implemented almost the same way right now, but different set of arguments:
```
def cond_functional_tensor_mode(pred, true_fn, false_fn, inputs):
    from torch._functorch.aot_autograd import from_fun, to_fun

    unwrapped_inputs = pytree.tree_map_only(FunctionalTensor, from_fun, inputs)
    unwrapped_pred = pytree.tree_map_only(FunctionalTensor, from_fun, pred)

    # We can rely on the Functionalize key to detect cond branches that modify the input
    functional_true = dispatch_functionalize(true_fn)
    functional_false = dispatch_functionalize(false_fn)

    with unset_functional_temporarily():
        cond_return = cond_op(
            unwrapped_pred, functional_true, functional_false, unwrapped_inputs
        )
    return pytree.tree_map_only(torch.Tensor, to_fun, cond_return)
```

Where we have to know that arguments 1 2 and 3 correspond to `pred` (need to unwrap) `true_fn` (need to functionalize) and `false_fn)` (need to functionalize).

It seems like a pain make a single FunctionalTensorMode function that's generic enough to work for all possible higher order op schemas. What did you have in mind? I could imagine something like this, although it seems a bit brittle since it might break if we add a new HigherOrderOp with a unique schema (was there something else you had in mind?)

```
def generic_functional_tensor_rule(
    higher_order_op: Callable,
    pre_fn_inputs: List[TensorLikeArg],
    fns: List[InnerFnArg],
    post_fn_inputs: List[TensorLikeArg]
):
    from torch._functorch.aot_autograd import from_fun, to_fun

    unwrapped_pre_args = pytree.tree_map_only(FunctionalTensor, from_fun, pre_fn_inputs)
    unwrapped_post_args = pytree.tree_map_only(FunctionalTensor, from_fun, post_fn_inputs)

    functional_fns = [dispatch_functionalize(f) for f in fns]

    with unset_functional_temporarily():
        out = higher_order_op(
            *unwrapped_pre_args, *functional_fns, *unwrapped_post_args
        )
    return pytree.tree_map_only(torch.Tensor, to_fun, out)
    
# to get the schemas to match up, we'd have to register them to map/cond like so:
def cond_functional_tensor_mode(pred, true_fn, false_fn, inputs):
    return generic_functional_tensor_rule(cond, [pred], [true_fn, false_fn], inputs)
def map_functional_tensor_mode(f, num_mapped, *args):
    return generic_functional_tensor_rule(map, [], [num_mapped], [num_mapped] + list(args))
```

zou3519 replied:
Does the following work?

```
def generic_functional_tensor_rule(
    higher_order_op,
    *args
):
    from torch._functorch.aot_autograd import from_fun, to_fun

    unwrapped_args = pytree.tree_map_only(FunctionalTensor, from_fun, args)

    with unset_functional_temporarily():
        out = higher_order_op(
            *unwrapped_args,
        )
    return pytree.tree_map_only(torch.Tensor, to_fun, out)
```

bdhirsh replied:
> I don't think there's a reason to functionalize the function args in the functional_tensor_rule

Oh I see. We'll actually need to wrap the inner functions in `dispatch_functionalize` in order for functionalizing subclasses to do the right thing on higher order ops (forcing functionalization to run underneath the subclass's `__torch_dispatch__`, **inside** of the inner_function that we're tracing out.

I guess we don't need this immediately, since none of the subclass that care about compile work with higher order ops anyway as you pointed out (since e.g. DTensor doesn't implement a rule for cond and friends).

One reason I can see doing this now (other than the fact that we'll need it at some point anyway) is just because it makes my mental model of functionalization simpler: During AOTAutograd, we will always use python functionalization to trace (both the outer graph, and any inner higher order ops). What do you think?

bdhirsh committed:
Update on "python functionalization: support higher order ops"

We now have two types of functionalization, C++ Functionalization (through the `Functionalize` dispatch key), and python functionalization (through the `FunctionalTensorMode` torch_dispatch mode).

This means that all higher order ops need custom functionalization rules for the python variant too. I added them here, as well as a helper function `dispatch_functionalize()` - equivalent to `torch.func.functionalize()`, except that it uses `FunctionalTensorMode`.

In theory we could have secretly switched `torch.func.functionalize` to use `FunctionalTensorMode`. This would be BC-breaking, though, since `FunctionalTensorMode` isn't composable with the other functorch transforms (the functorch layer-mode stack doesn't know how to re-order torch_dispatch modes arbitrarily).




[ghstack-poisoned]

bdhirsh committed:
Update on "python functionalization: support higher order ops"

We now have two types of functionalization, C++ Functionalization (through the `Functionalize` dispatch key), and python functionalization (through the `FunctionalTensorMode` torch_dispatch mode).

This means that all higher order ops need custom functionalization rules for the python variant too. I added them here, as well as a helper function `dispatch_functionalize()` - equivalent to `torch.func.functionalize()`, except that it uses `FunctionalTensorMode`.

In theory we could have secretly switched `torch.func.functionalize` to use `FunctionalTensorMode`. This would be BC-breaking, though, since `FunctionalTensorMode` isn't composable with the other functorch transforms (the functorch layer-mode stack doesn't know how to re-order torch_dispatch modes arbitrarily).




[ghstack-poisoned]

bdhirsh commented on functorch/experimental/_map.py:
```diff
@@ -289,6 +331,9 @@ def _stack_pytree(pytrees):
             # therefore we need to deal with None output.
             stacked_out.append(None)
         else:
+            import pdb
```
whoops removing

bdhirsh committed:
Update on "python functionalization: support higher order ops"

We now have two types of functionalization, C++ Functionalization (through the `Functionalize` dispatch key), and python functionalization (through the `FunctionalTensorMode` torch_dispatch mode).

This means that all higher order ops need custom functionalization rules for the python variant too. I added them here, as well as a helper function `dispatch_functionalize()` - equivalent to `torch.func.functionalize()`, except that it uses `FunctionalTensorMode`.

In theory we could have secretly switched `torch.func.functionalize` to use `FunctionalTensorMode`. This would be BC-breaking, though, since `FunctionalTensorMode` isn't composable with the other functorch transforms (the functorch layer-mode stack doesn't know how to re-order torch_dispatch modes arbitrarily).




[ghstack-poisoned]

bdhirsh committed:
Update on "python functionalization: support higher order ops"

We now have two types of functionalization, C++ Functionalization (through the `Functionalize` dispatch key), and python functionalization (through the `FunctionalTensorMode` torch_dispatch mode).

This means that all higher order ops need custom functionalization rules for the python variant too. I added them here, as well as a helper function `dispatch_functionalize()` - equivalent to `torch.func.functionalize()`, except that it uses `FunctionalTensorMode`.

In theory we could have secretly switched `torch.func.functionalize` to use `FunctionalTensorMode`. This would be BC-breaking, though, since `FunctionalTensorMode` isn't composable with the other functorch transforms (the functorch layer-mode stack doesn't know how to re-order torch_dispatch modes arbitrarily).




[ghstack-poisoned]

bdhirsh committed:
Update on "python functionalization: support higher order ops"

We now have two types of functionalization, C++ Functionalization (through the `Functionalize` dispatch key), and python functionalization (through the `FunctionalTensorMode` torch_dispatch mode).

This means that all higher order ops need custom functionalization rules for the python variant too. I added them here, as well as a helper function `dispatch_functionalize()` - equivalent to `torch.func.functionalize()`, except that it uses `FunctionalTensorMode`.

In theory we could have secretly switched `torch.func.functionalize` to use `FunctionalTensorMode`. This would be BC-breaking, though, since `FunctionalTensorMode` isn't composable with the other functorch transforms (the functorch layer-mode stack doesn't know how to re-order torch_dispatch modes arbitrarily).




[ghstack-poisoned]

bdhirsh committed:
Update on "python functionalization: support higher order ops"

We now have two types of functionalization, C++ Functionalization (through the `Functionalize` dispatch key), and python functionalization (through the `FunctionalTensorMode` torch_dispatch mode).

This means that all higher order ops need custom functionalization rules for the python variant too. I added them here, as well as a helper function `dispatch_functionalize()` - equivalent to `torch.func.functionalize()`, except that it uses `FunctionalTensorMode`.

In theory we could have secretly switched `torch.func.functionalize` to use `FunctionalTensorMode`. This would be BC-breaking, though, since `FunctionalTensorMode` isn't composable with the other functorch transforms (the functorch layer-mode stack doesn't know how to re-order torch_dispatch modes arbitrarily).




[ghstack-poisoned]

bdhirsh committed:
Update on "python functionalization: support higher order ops"

We now have two types of functionalization, C++ Functionalization (through the `Functionalize` dispatch key), and python functionalization (through the `FunctionalTensorMode` torch_dispatch mode).

This means that all higher order ops need custom functionalization rules for the python variant too. I added them here, as well as a helper function `dispatch_functionalize()` - equivalent to `torch.func.functionalize()`, except that it uses `FunctionalTensorMode`.

In theory we could have secretly switched `torch.func.functionalize` to use `FunctionalTensorMode`. This would be BC-breaking, though, since `FunctionalTensorMode` isn't composable with the other functorch transforms (the functorch layer-mode stack doesn't know how to re-order torch_dispatch modes arbitrarily).




[ghstack-poisoned]

bdhirsh committed:
Update on "python functionalization: support higher order ops"

We now have two types of functionalization, C++ Functionalization (through the `Functionalize` dispatch key), and python functionalization (through the `FunctionalTensorMode` torch_dispatch mode).

This means that all higher order ops need custom functionalization rules for the python variant too. I added them here, as well as a helper function `dispatch_functionalize()` - equivalent to `torch.func.functionalize()`, except that it uses `FunctionalTensorMode`.

In theory we could have secretly switched `torch.func.functionalize` to use `FunctionalTensorMode`. This would be BC-breaking, though, since `FunctionalTensorMode` isn't composable with the other functorch transforms (the functorch layer-mode stack doesn't know how to re-order torch_dispatch modes arbitrarily).




[ghstack-poisoned]

zou3519 commented on torch/_ops.py:
```diff
@@ -127,6 +127,110 @@ def inner(fn):
 
+    # Registers an implementation to all **3** variants of functionalization that we have:
+    # - DispatchKey.Functionalize
+    # - functorch.TransformType.FuncTorchFunctionalize
```
nit: this is just functorch.TransformType.Functionalize, right?

zou3519 commented on torch/_ops.py:
```diff
@@ -127,6 +127,110 @@ def inner(fn):
+                for k, v in d.items():
+                    setattr(self, k, v)
+
+        functionalize_dk_methods = {
+            "unwrap": lambda args: _unwrap_all_tensors_from_functional(
+                args, reapply_views=_reapply_views()
+            ),
+            "wrap": lambda args: _wrap_all_tensors_to_functional(args, level=0),
+            "wrap_function": lambda f: torch.func.functionalize(
+                f, remove="mutations_and_views" if _reapply_views() else "mutations"
+            ),
+            "redispatch": lambda: torch._C._ExcludeDispatchKeyGuard(
+                torch._C.DispatchKeySet(torch._C.DispatchKey.Functionalize)
+            ),
+        }
+
```
For code readability: could we author these dicts as classes instead of dicts?  That is, there is some BaseFunctionalizeAPI class with four methods: wrap, unwrap, wrap_function, redispatch, then FunctorchFunctionalizeAPI, PythonFunctionalizeAPI, and CppFunctionalizeAPI can subclass from it

Also, at the risk of complicating your PR stack: something that might help is moving HigherOrderOp out of torch/_ops.py because every global member of torch/_ops.py becomes torch.ops.blah (and therefore public API)

bdhirsh replied:
> For code readability: could we author these dicts as classes instead of dicts? 

Sure - I originally do the dict-to-class approach just to minimize boilerplate, but happy to make them actual classes for readability.

> Also, at the risk of complicating your PR stack: something that might help is moving HigherOrderOp out of torch/_ops.py because every global member of torch/_ops.py becomes torch.ops.blah (and therefore public API)

Hmm. This is totally fair. Just to minimize the changes in the PR, are you ok with me authoring the new classes in `torch/_subclasses/functional_tensor.py` and importing them into here? I agree that moving `HigherOrderOperator` out of `torch/_ops.py` feels like the right thing though.

zou3519 replied:
> Hmm. This is totally fair. Just to minimize the changes in the PR, are you ok with me authoring the new classes in torch/_subclasses/functional_tensor.py and importing them into here? I agree that moving HigherOrderOperator out of torch/_ops.py feels like the right thing though.

Yes, as long as we confirm that importing them into torch._ops doesn't make them public. The torch._ops -> torch.ops thing makes me nervous

bdhirsh replied:
agreed - i'll confirm that this is the case

zou3519 commented on torch/_ops.py:
```diff
@@ -127,6 +127,110 @@ def inner(fn):
+    # Registers an implementation to all **3** variants of functionalization that we have:
+    # - DispatchKey.Functionalize
+    # - functorch.TransformType.FuncTorchFunctionalize
+    # - FunctionalTensorMode
+    def py_functionalize_impl(self):
```
Could you add an example of how to use it in the comment?

bdhirsh replied:
will do

zou3519 commented on functorch/experimental/_map.py:
```diff
@@ -320,66 +356,25 @@ def map_fake_tensor_mode(mode, f, num_mapped, *args):
-@map_impl.py_impl(DispatchKey.Functionalize)
-def map_func(f, num_mapped, *args):
-    reapply_views = torch._C._functionalization_reapply_views_tls()
+@map_impl.py_functionalize_impl()
```
nit: Does py_functionalize_impl take in any args? If not then we should make it so that the () aren't necessary

bdhirsh replied:
It does not - agreed (let me make sure that this works)

zou3519 commented on functorch/experimental/_map.py:
```diff
@@ -320,66 +356,25 @@ def map_fake_tensor_mode(mode, f, num_mapped, *args):
-    mode = "mutations_and_views" if reapply_views else "mutations"
+    unwrapped_xs = ctx.unwrap(xs)
+    unwrapped_args = ctx.unwrap(pos_args)
+    wrapped_fn = ctx.wrap_function(f)
 
-    with _ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.Functionalize)):
-        functional_map_fn = functionalize(f, remove=mode)
+    with ctx.redispatch():
```
nit: suggesting some more descriptive names:
- unwrap -> ctx.unwrap_tensors
- wrap -> ctx.wrap_tensors
- wrap_function -> ctx.functionalize (the functionalization API)
- redispatch() -> redispatch_to_next()

bdhirsh replied:
I like all these better, thanks - will change!

zou3519 commented on functorch/experimental/_map.py:
```diff
@@ -25,6 +27,23 @@
 from torch.multiprocessing.reductions import StorageWeakRef
 
 
+# TODO: pull these helpers from AOTAutograd later
```
Do you mean in this PR or later?

zou3519 commented on functorch/experimental/_map.py:
```diff
@@ -65,10 +84,10 @@ def create_fw_bw_graph(f, num_mapped_args, *args):
     # Instead, it will create _tensor_constant as output.
 
-    with suspend_functionalization():
+    with suspend_functionalization(), unset_functional_temporarily():
```
IMO FunctionalTensorMode should automatically `unset_functional_temporarily` to be analogous to functorch.functionalize. Would there be problems if we did that?

bdhirsh replied:
Hmm I don't think I follow - what behavior in `functorch.functionalize` is that analogous to?

(also - I'm probably missing something, but I'm not sure how that's related to this line. If I understand correctly, we want to ahead-of-time trace out a fw and bw graph here, which we will then invoke from the autograd rule, and then functionalize later, which is why we're ok with turning off functionalization now. But doesn't that mean that we need to explicitly disable functionalization here? (both the C++ variant by disabling the `Functionalize` dispatch key, and the python variant by temporarily disabling any ambient `FunctionalTensorMode` that might be active).

zou3519 replied:
sorry, I misread unset_functional_temporarily() as "turn on the functionalization key", we are actually disabling it to be in-line with suspend_functionalization() so that makes sense to me

zou3519 commented on torch/_functorch/eager_transforms.py:
```diff
@@ -1291,21 +1292,29 @@ def grad_impl(func: Callable, argnums: argnums_t, has_aux: bool, args, kwargs):
     return grad
 
-def _maybe_wrap_functional_tensor(maybe_tensor, level):
+def _maybe_wrap_functional_tensor(maybe_tensor, level, *, _python_functionalize: bool = False):
```
Why did we need to change these APIs?

zou3519 commented on torch/_ops.py:
```diff
@@ -127,6 +127,110 @@ def inner(fn):
+                ctx = _DictToClass(functionalize_functorch_methods)
+                return fn(ctx, *args, **kwargs)
+
+            assert FunctionalTensorMode not in self.python_key_mode_table
+            self.python_key_mode_table[
+                FunctionalTensorMode
+            ] = functionalize_dispatch_mode_fn
+
+            assert (
+                torch._C._functorch.TransformType.Functionalize
+                not in self.functorch_table
+            )
+            self.functorch_table[
+                torch._C._functorch.TransformType.Functionalize
+            ] = functionalize_functorch_fn
+
+            if torch._C.DispatchKey.Functionalize in self.py_kernels:
+                raise RuntimeError(
+                    f"Trying to override a python impl for DispatchKey.Functionalize on operator {self.name()}"
+                )
+            self.py_kernels[torch._C.DispatchKey.Functionalize] = functionalize_dk_fn
+
+            self._dispatch_cache.clear()
+            return fn
```
Could we call py_functorch_impl and py_impl instead of manually assigning to the tables?

bdhirsh replied:
I think so, yep. Will do

zou3519 commented on test/functorch/test_control_flow.py:
```diff
@@ -5,13 +5,43 @@
 from torch.testing._internal.common_quantization import skipIfNoDynamoSupport
+from torch._subclasses.functional_tensor import FunctionalTensor
+
+# TODO: pull these helpers from AOTAutograd later
```
Needs some deduping

bdhirsh replied:
Yep - I'm definitely cluttering up the tests will multiple copies of these floating around.

Are you ok with me doing the cleanup in https://github.com/pytorch/pytorch/pull/106406? (this feels like less churn to me, since I think that long term it's reasonable for AOTAutograd to be the source of truth for these API's anyway)

zou3519 replied:
Yeah, that sgtm

zou3519 commented on test/functorch/test_control_flow.py:
```diff
@@ -1429,9 +1463,11 @@ def f_wrapper(func):
-                    func_args = pytree.tree_map(to_fun, args)
-                    func_kwargs = pytree.tree_map(to_fun, kwargs)
-                    return pytree.tree_map(from_fun, func(*func_args, **func_kwargs))
+                    func_args = pytree.tree_map(
+                        lambda x: to_fun_old(x) if isinstance(x, torch.Tensor) else x, args)
+                    func_kwargs = pytree.tree_map(
```
TBH if we do the refactor into FunctorchFunctionalizeAPI, CppFunctionalizeAPI, PythonFunctionalizeAPI, we could use those as the sources of truth and just call CppFunctionalizeAPI.unwrap_tensors(...)

zou3519 commented on functorch/experimental/_map.py:
```diff
@@ -320,66 +356,25 @@ def map_fake_tensor_mode(mode, f, num_mapped, *args):
-        pos_args, reapply_views=reapply_views
-    )
-    mode = "mutations_and_views" if reapply_views else "mutations"
+    unwrapped_xs = ctx.unwrap(xs)
+    unwrapped_args = ctx.unwrap(pos_args)
```
Should unwrap/wrap unwrap a single Tensor or do the tree_map_only(torch.Tensor)?

bdhirsh replied:
All 3 of the functionalization rules that we have end up doing a tree_map, so it felt generic enough that I abstracted it away into the `ctx` instead of making the higher order op author spell it out.

But making the higher order op author do it also seems reasonable, let me know if you have a preference

zou3519 replied:
No real preference, I agree that everywhere we  use it it seems to be used with a tree_map_only so that sounds reasonable to me

zou3519 reviewed (commented):
I need to go read the map() code in some more detail. I don't remember what we're doing there. But everything else seems reasonable. My only comments are suggestions to do some refactoring and to repeat ourselves less

zou3519 commented on functorch/experimental/_map.py:
```diff
@@ -80,17 +99,27 @@ def from_fun(t):
                         # but we want to avoid it so we clone a non-functional version
-                        maybe_unfunc_t = torch._functorch.aot_autograd.from_fun(t)
+                        maybe_unfunc_t = t
+                        if isinstance(t, FunctionalTensor):
+                            torch._sync(t)
+                            maybe_unfunc_t = from_fun(t)
+                        elif torch._is_functional_tensor(t):
+                            # need to handle both types of functionalization here:
+                            # these are the tensors that came from the user,
+                            # which could be either FunctionalTensorWrapper or FunctionalTensor
+                            torch._sync(t)
+                            maybe_unfunc_t = torch._from_functional_tensor(t)
```
I'm confused about why we need to handle both FunctionalTensor here and C++ Functionalization. The way we actually hit this code is through AOTAutograd. I expect us to move AOTAutograd to Python Functionalization (not sure where this is in your stack). So we shouldn't need to have a branch for C++ Functionalization anymore

bdhirsh replied:
Ah ok - I wasn't sure if this code was only meant to be used in AOTAutograd, or also might be used more generically, so I played it safe and just made it work with both flavors.

What do you think of leaving as-is in this PR, and me removing the C++ functionalization handling in the PR where I actually flip the switch on python functionalization in AOTAutograd?

I mainly tried to set the PR's up in that way so that there was as little code as possible in the "turn on python functionalization in AOTAutograd", since that PR carries slightly higher revert risk.

zou3519 replied:
Sounds fine

zou3519 reviewed (commented):
I remembered what the map implementation does now, left a question about it

bdhirsh committed:
Update on "python functionalization: support higher order ops"

We now have two types of functionalization, C++ Functionalization (through the `Functionalize` dispatch key), and python functionalization (through the `FunctionalTensorMode` torch_dispatch mode).

This means that all higher order ops need custom functionalization rules for the python variant too. I added them here, as well as a helper function `dispatch_functionalize()` - equivalent to `torch.func.functionalize()`, except that it uses `FunctionalTensorMode`.

In theory we could have secretly switched `torch.func.functionalize` to use `FunctionalTensorMode`. This would be BC-breaking, though, since `FunctionalTensorMode` isn't composable with the other functorch transforms (the functorch layer-mode stack doesn't know how to re-order torch_dispatch modes arbitrarily).




[ghstack-poisoned]

bdhirsh committed:
Update on "python functionalization: support higher order ops"

We now have two types of functionalization, C++ Functionalization (through the `Functionalize` dispatch key), and python functionalization (through the `FunctionalTensorMode` torch_dispatch mode).

This means that all higher order ops need custom functionalization rules for the python variant too. I added them here, as well as a helper function `dispatch_functionalize()` - equivalent to `torch.func.functionalize()`, except that it uses `FunctionalTensorMode`.

In theory we could have secretly switched `torch.func.functionalize` to use `FunctionalTensorMode`. This would be BC-breaking, though, since `FunctionalTensorMode` isn't composable with the other functorch transforms (the functorch layer-mode stack doesn't know how to re-order torch_dispatch modes arbitrarily).




[ghstack-poisoned]

zou3519 commented on torch/_export/wrappers.py:
```diff
@@ -38,17 +37,14 @@ def export_tracepoint_fake_tensor_mode(mode, *args, **kwargs):
-    with _ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.Functionalize)):
-        return _export_tracepoint(*unwrapped_args, **unwrapped_kwargs)
+@_export_tracepoint.py_functionalize_impl
+def export_tracepoint_functional_(ctx, *args, **kwargs):
```
why the trailing _?

bdhirsh replied:
poor rename, will fix

zou3519 commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -254,3 +256,134 @@ def unset_functional_temporarily():
+    return inner
+
+
+class BaseFunctionalizeAPI(ABC):
+    @abstractmethod
+    def wrap_tensors(self, args):
+        pass
+
+    @abstractmethod
+    def unwrap_tensors(self, args):
+        pass
+
+    @abstractmethod
+    def functionalize(self, inner_f):
+        pass
+
+    @abstractmethod
+    def redispatch_to_next(self):
+        pass
+
```
Can we type these? In particular it's important for readability that redispatch_to_next gets typed as returning a context manager

zou3519 reviewed (approved):
LGTM, minor comments

bdhirsh committed:
Update on "python functionalization: support higher order ops"

We now have two types of functionalization, C++ Functionalization (through the `Functionalize` dispatch key), and python functionalization (through the `FunctionalTensorMode` torch_dispatch mode).

This means that all higher order ops need custom functionalization rules for the python variant too. I added them here, as well as a helper function `dispatch_functionalize()` - equivalent to `torch.func.functionalize()`, except that it uses `FunctionalTensorMode`.

In theory we could have secretly switched `torch.func.functionalize` to use `FunctionalTensorMode`. This would be BC-breaking, though, since `FunctionalTensorMode` isn't composable with the other functorch transforms (the functorch layer-mode stack doesn't know how to re-order torch_dispatch modes arbitrarily).




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "python functionalization: support higher order ops"

We now have two types of functionalization, C++ Functionalization (through the `Functionalize` dispatch key), and python functionalization (through the `FunctionalTensorMode` torch_dispatch mode).

This means that all higher order ops need custom functionalization rules for the python variant too. I added them here, as well as a helper function `dispatch_functionalize()` - equivalent to `torch.func.functionalize()`, except that it uses `FunctionalTensorMode`.

In theory we could have secretly switched `torch.func.functionalize` to use `FunctionalTensorMode`. This would be BC-breaking, though, since `FunctionalTensorMode` isn't composable with the other functorch transforms (the functorch layer-mode stack doesn't know how to re-order torch_dispatch modes arbitrarily).




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
